File: Fiber_Attraction_ABP/Fiber_Attraction_ABP.cpp
#include <iostream>
#include <fstream>
#include <cmath>
#include <omp.h>
#include <random>
#include <vector>

//*************Set Parameters**************
const int NUMSIMS = 1;  // Number of simulations
const int STEP = 10000; // Number of movement steps
const int TICK = 10;    // Number of steps between each save file

const double LENGTH = 1;             // The length of box
const double DENSITY = 0.3;          // The density of system
const double RADIUS = 0.02;          // The radius of particles
const double DELTATIME = 1;          // particles[i].position[0] += DELTATIME * particles[i].velocity[0];
const double V = 0.001;              //Initial velocity
const double DIFFCOEFF = 1;          // The diffusion coefficient
const double KCON = RADIUS * 0.1;    // The coefficient for contact force
const double KECM = RADIUS * 0.0001; // The coefficient for ECM force
double ANGLE = 1;                    // For ECM force

const double DIFFCOEFF2 = RADIUS * 0.00001; // The diffusion coefficient for random force
//const double EPS = 0.0000000002;            // A samll number for preventing division divergence
const double FSELF = RADIUS * 0.1; // The self propulsion force

const double PI = 3.1415926;
const double AREAOFBOX = LENGTH * LENGTH;
const int NUMBER = int(AREAOFBOX * DENSITY / (PI * RADIUS * RADIUS));

//Add environment effect
const int NUMGRID = 5; // Number of gird
const int MAXPERGRID = int(LENGTH / NUMGRID / RADIUS / 2 + 1) * int(LENGTH / NUMGRID / RADIUS / 2 + 1);
//double Grid[NUMGRID][NUMGRID][2];
struct Grid
{
    double direction[2];
    int particlesInGrid[MAXPERGRID];
    int numOfParticles;
    const double initDirectionX = 1;
    const double initDirectionY = 0;
};
const double fiberChangeCoeff = 0.5;  //for direction of fiber
const double fiberRecoverCoeff = 0.5; // for direction of fiber recover
const double fiberEffectCoeff = 0.5;  //for velocity of particles

//*************Set Parameters**************

//********Set Random Numbers*********
std::random_device rd;
std::default_random_engine gen(rd());
std::normal_distribution<double> norm_distribution_angle(0, PI / 32);
std::uniform_real_distribution<float> uniform_distribution(0.0, 1.0);
//std::uniform_real_distribution<float> uniform_distribution2(-1.0, 1.0);
//********Set Random Numbers*********

struct Particle
{
    double position[2];
    double velocity[2];
    double force[2];
    double correctMove[2];
    int labelofgrid[2];
};

double boundaryDisplacementX(Particle a, Particle b)
{
    double deltax = a.position[0] - b.position[0];

    if (std::abs(deltax) > LENGTH / 2)
    {
        if (deltax > 0)
            deltax -= LENGTH;
        else
            deltax += LENGTH;
    }
    return deltax;
}

double boundaryDisplacementY(Particle a, Particle b)
{
    double deltay = a.position[1] - b.position[1];

    if (std::abs(deltay) > LENGTH / 2)
    {
        if (deltay > 0)
            deltay -= LENGTH;
        else
            deltay += LENGTH;
    }
    return deltay;
}

double boundaryDistance(Particle a, Particle b)
{
    double deltax = boundaryDisplacementX(a, b);
    double deltay = boundaryDisplacementY(a, b);

    return std::sqrt(deltax * deltax + deltay * deltay);
}

void boundaryCondition(Particle &a)
{
    if (a.position[0] >= LENGTH)
        a.position[0] = a.position[0] - LENGTH;
    if (a.position[0] < 0)
        a.position[0] = a.position[0] + LENGTH;
    if (a.position[1] >= LENGTH)
        a.position[1] = a.position[1] - LENGTH;
    if (a.position[1] < 0)
        a.position[1] = a.position[1] + LENGTH;
}

bool isOverlap(Particle a, Particle b)
{
    return (boundaryDistance(a, b) < 2 * RADIUS);
}

bool isOppositeMove(Particle a, Particle b)
{
    double d = boundaryDistance(a, b);
    double dv = std::sqrt(a.velocity[0] * a.velocity[0] + a.velocity[1] * a.velocity[1]);
    double deltax, deltay, deltax1, deltay1;
    double cosTheta;
    deltax = boundaryDisplacementX(a, b);
    deltay = boundaryDisplacementY(a, b);

    cosTheta = (-deltax * a.velocity[0] + (-deltay) * a.velocity[1]) / (d * dv);

    return (cosTheta >= cos(ANGLE * PI / 180) and (a.velocity[0] * b.velocity[0] + a.velocity[1] * b.velocity[1] < 0));
}

void initializationParticles(Particle particles[])
{
    int i = 0;
    std::cout << "inti" << std::endl;
    while (i < NUMBER)
    {
        particles[i].position[0] = LENGTH * (uniform_distribution(gen));
        particles[i].position[1] = LENGTH * (uniform_distribution(gen));

        particles[i].labelofgrid[0] = std::min(int(floor(particles[i].position[0] / (LENGTH / NUMGRID))), NUMGRID - 1);
        particles[i].labelofgrid[1] = std::min(int(floor(particles[i].position[1] / (LENGTH / NUMGRID))), NUMGRID - 1);

        bool hasOverlap = false;
        for (int j = 0; j < i; j++)
        {
            if (i != j && isOverlap(particles[i], particles[j]))
            {
                hasOverlap = true;
                break;
            }
        }
        if (!hasOverlap)
        {
            i++;
            //std::cout << "Add: " << i << " out of " << NUMBER << std::endl;
        }
    }
    int ii;
#pragma omp parallel for private(ii)
    for (ii = 0; ii < NUMBER; ii++)
    {
        particles[ii].velocity[0] = V * LENGTH * (uniform_distribution(gen) - 0.5);
        particles[ii].velocity[1] = V * LENGTH * (uniform_distribution(gen) - 0.5);
        particles[ii].force[0] = 0;
        particles[ii].force[1] = 0;
        particles[ii].correctMove[0] = 0;
        particles[ii].correctMove[1] = 0;
    }
}

void initializationGrids(Grid grids[][NUMGRID])
{
    for (int i = 0; i < NUMGRID; i++)
    {
        for (int j = 0; j < NUMGRID; j++)
        {
            grids[i][j].direction[0] = 1; //uniform_distribution2(gen);
            grids[i][j].direction[1] = 0; //uniform_distribution2(gen);
            grids[i][j].numOfParticles = 0;
        }
    }
}

void forceContactECM(Particle particles[])
{
    int i, j;
#pragma omp parallel for private(j)
    for (i = 0; i < NUMBER; i++)
    {
        for (j = 0; j < NUMBER; j++)
        {
            // if (isOverlap(particles[i], particles[j]) & (i != j))
            // {
            //     double d = boundaryDistance(particles[i], particles[j]);
            //     double deltax, deltay;
            //     deltax = boundaryDisplacementX(particles[i], particles[j]);
            //     deltay = boundaryDisplacementY(particles[i], particles[j]);

            //     particles[i].force[0] += DIFFCOEFF * KCON * std::abs(2 * RADIUS - d) * deltax / d;
            //     particles[i].force[1] += DIFFCOEFF * KCON * std::abs(2 * RADIUS - d) * deltay / d;
            // }
            //  else

            if (isOppositeMove(particles[i], particles[j]) & (i != j))
            {
                double d = boundaryDistance(particles[i], particles[j]);
                double deltax, deltay;
                deltax = boundaryDisplacementX(particles[i], particles[j]);
                deltay = boundaryDisplacementY(particles[i], particles[j]);

                particles[i].force[0] += DIFFCOEFF * (KECM / d * (-deltax / d));
                particles[i].force[1] += DIFFCOEFF * (KECM / d * (-deltay / d));
            }
        }
    }
}

// void forceSelf(Particle particles[NUMBER])
// {
//     int i;
// #pragma omp parallel for private(i)
//     for (i = 0; i < NUMBER; i++)
//     {
//         double v = sqrt(particles[i].velocity[0] * particles[i].velocity[0] + particles[i].velocity[1] * particles[i].velocity[1]);
//         particles[i].force[0] += DIFFCOEFF * FSELF * particles[i].velocity[0] / v;
//         particles[i].force[1] += DIFFCOEFF * FSELF * particles[i].velocity[1] / v;
//     }
// }

// void forceRandom(Particle particles[NUMBER])
// {
//     int i;
// #pragma omp parallel for private(i)
//     for (i = 0; i < NUMBER; i++)
//     {
//         particles[i].force[0] += DIFFCOEFF2 * LENGTH * (uniform_distribution(gen) - 0.5);
//         particles[i].force[1] += DIFFCOEFF2 * LENGTH * (uniform_distribution(gen) - 0.5);
//     }
// }

double rotateAtoB(double Ax, double Ay, double Bx, double By, double coeff)
{
    double Aangle = atan2(Ay, Ax);
    double Bangle = atan2(By, Bx);

    double diff = Bangle - Aangle;
    double sign = (diff < 0) ? -1.0 : 1.0;

    while (std::abs(diff) > PI / 2)
    {
        diff -= PI * sign;
    }
    return diff * coeff;
}

void fiberAffected(Particle particles[], Grid grids[][NUMGRID])
{
    int i;
#pragma omp parallel for private(i)
    for (i = 0; i < NUMBER; i++)
    {
        particles[i].labelofgrid[0] = std::min(int(floor(particles[i].position[0] / (LENGTH / NUMGRID))), NUMGRID - 1);
        particles[i].labelofgrid[1] = std::min(int(floor(particles[i].position[1] / (LENGTH / NUMGRID))), NUMGRID - 1);

        int n = grids[particles[i].labelofgrid[0]][particles[i].labelofgrid[1]].numOfParticles;
        grids[particles[i].labelofgrid[0]][particles[i].labelofgrid[1]].particlesInGrid[n] = i;
        grids[particles[i].labelofgrid[0]][particles[i].labelofgrid[1]].numOfParticles += 1;
    }
    int ii, j;
    //#pragma omp parallel for private(ii)
    for (ii = 0; ii < NUMGRID; ii++)
    {
        for (j = 0; j < NUMGRID; j++)
        {
            double fiberX = grids[ii][j].direction[0]; //The x of direction of fiber
            double fiberY = grids[ii][j].direction[1]; //The y of direction of fiber

            double fiber = std::sqrt(fiberX * fiberX + fiberY * fiberY);
            double fiberAngle = atan2(grids[ii][j].direction[1], grids[ii][j].direction[0]);
            double rotateAngle = 0;
            if (grids[ii][j].numOfParticles >= 2)
            {
                for (int p1 = 0; p1 < grids[ii][j].numOfParticles; p1++)
                {
                    for (int p2 = p1 + 1; p2 < grids[ii][j].numOfParticles; p2++)
                    {
                        int a, b;
                        a = grids[ii][j].particlesInGrid[p1];
                        if (isOppositeMove(particles[a], particles[b]))
                        {
                            rotateAngle += rotateAtoB(fiberX, fiberY, particles[a].position[0] - particles[b].position[0], particles[a].position[1] - particles[b].position[1], fiberChangeCoeff);
                        }
                    }
                }
                grids[ii][j].direction[0] = fiber * cos(fiberAngle + rotateAngle);
                grids[ii][j].direction[1] = fiber * sin(fiberAngle + rotateAngle);
            }
            else
            {
                double initfiberX = grids[ii][j].initDirectionX;
                double initfiberY = grids[ii][j].initDirectionY;
                double rotateAngle = rotateAtoB(fiberX, fiberY, initfiberX, initfiberY, fiberRecoverCoeff);
                grids[ii][j].direction[0] = fiber * cos(fiberAngle + rotateAngle);
                grids[ii][j].direction[1] = fiber * sin(fiberAngle + rotateAngle);
            }
            grids[ii][j].numOfParticles = 0;
        }
    }
}

void move(Particle particles[], Grid grids[][NUMGRID])
{
    /*
    Move all particles. Update the positions of particles 
    according to the current velocities.Velocities are updated 
    by Gaussian random walk. Gaussian distribution is set by
    global normal distribution generator.
    */
    int i;
#pragma omp parallel for private(i)
    for (i = 0; i < NUMBER; i++)
    {

        particles[i].labelofgrid[0] = std::min(int(floor(particles[i].position[0] / (LENGTH / NUMGRID))), NUMGRID - 1);
        particles[i].labelofgrid[1] = std::min(int(floor(particles[i].position[1] / (LENGTH / NUMGRID))), NUMGRID - 1);
        double v = std::sqrt(particles[i].velocity[0] * particles[i].velocity[0] + particles[i].velocity[1] * particles[i].velocity[1]);
        double velocityAngle;
        //Add force
        particles[i].velocity[0] += particles[i].force[0];
        particles[i].velocity[1] += particles[i].force[1];
        velocityAngle = atan2(particles[i].velocity[1], particles[i].velocity[0]);
        //particles[i].velocity[0] = v * cos(velocityAngle);
        //particles[i].velocity[1] = v * sin(velocityAngle);

        //environment effect

        double fiberX = grids[particles[i].labelofgrid[0]][particles[i].labelofgrid[1]].direction[0]; //The x of direction of fiber
        double fiberY = grids[particles[i].labelofgrid[0]][particles[i].labelofgrid[1]].direction[1]; //The y of direction of fiber
        //double fiberAngle = getAngle(fiberX, fiberY);

        velocityAngle += rotateAtoB(particles[i].velocity[0], particles[i].velocity[1], fiberX, fiberY, fiberEffectCoeff);
        //rotation diffusion
        velocityAngle += norm_distribution_angle(gen);

        particles[i].velocity[0] = v * cos(velocityAngle);
        particles[i].velocity[1] = v * sin(velocityAngle);
        ////////////////////

        //Update positions
        particles[i].position[0] += DELTATIME * particles[i].velocity[0];
        particles[i].position[1] += DELTATIME * particles[i].velocity[1];

        // Correct the position according to periodic boundary condition.
        boundaryCondition(particles[i]);
    }
}
void correctOverlap(Particle &a, Particle &b)
{
    /*
    Offset overlap between two particles.If two particles overlap, 
    move each one-half the overlap distance along their center-to-center axis.
    */
    double d = boundaryDistance(a, b);
    double deltax = boundaryDisplacementX(a, b);
    double deltay = boundaryDisplacementY(a, b);
    double deltaD = 2 * RADIUS - d;

    a.correctMove[0] += 0.5 * deltaD * deltax / d;
    a.correctMove[1] += 0.5 * deltaD * deltay / d;
    //b.position[0] += 0.5 * deltaD * (-deltax) / d;
    //b.position[1] += 0.5 * deltaD * (-deltay) / d;
}
void recordCorrectOverlap(Particle particles[])
{
    int i, j;
#pragma omp parallel for private(j)
    for (i = 0; i < NUMBER; i++)
    {
        for (j = 0; j < NUMBER; j++)
        {
            if (j != i)
            {
                if (isOverlap(particles[i], particles[j]))
                {
                    correctOverlap(particles[i], particles[j]);
                }
            }
        }
    }
}

void moveCorrectOverlap(Particle particles[])
{
    int i;
#pragma omp parallel for private(i)
    for (i = 0; i < NUMBER; i++)
    {
        particles[i].position[0] += particles[i].correctMove[0];
        particles[i].position[1] += particles[i].correctMove[1];
        particles[i].correctMove[0] = 0;
        particles[i].correctMove[1] = 0;
        boundaryCondition(particles[i]);
    }
}
void simulation(int simID)
{
    /*
    Run a complete simulation from start. simID labels the save detination for data.
    After initialization, particles move for a number of steps specified by the global
    constant STEP. Motion data is saved for every TICK steps.
    */
    Particle particles[NUMBER];
    Grid grids[NUMGRID][NUMGRID];
    initializationParticles(particles);
    initializationGrids(grids);
    std::ofstream fout1("data" + std::to_string(simID) + "/position0.txt");
    for (int i = 0; i < NUMBER; i++)
    {
        fout1 << particles[i].position[0] << " " << particles[i].position[1] << " " << particles[i].velocity[0] << " " << particles[i].velocity[1] << std::endl;
    }
    fout1.close();

    std::ofstream fout3("data" + std::to_string(simID) + "/fiberPosition0.txt");

    for (int i = 0; i < NUMGRID; i++)
        for (int j = 0; j < NUMGRID; j++)
        {
            fout3 << grids[i][j].direction[0] << " " << grids[i][j].direction[1] << std::endl;
        }
    fout3.close();

    for (int step = 0; step < STEP; step++)
    {
        for (int i = 0; i < NUMBER; i++)
        {
            particles[i].force[0] = 0;
            particles[i].force[1] = 0;
        }
        //forceContactECM(particles);
        //forceSelf(particles);
        //forceRandom(particles);
        move(particles, grids);
        recordCorrectOverlap(particles);
        moveCorrectOverlap(particles);
        fiberAffected(particles, grids);
        if ((step + 1) % TICK == 0)
        {
            //std::cout << "Step " << step + 1 << " out of " << STEP << std::endl;
            const std::string fileName = "data" + std::to_string(simID) + "/position" + std::to_string(step + 1) + ".txt";
            std::ofstream fout(fileName);
            for (int i = 0; i < NUMBER; i++)
            {
                fout << particles[i].position[0] << " " << particles[i].position[1] << " " << particles[i].velocity[0] << " " << particles[i].velocity[1] << std::endl;
            }
            fout.close();

            const std::string fileName2 = "data" + std::to_string(simID) + "/fiberPosition" + std::to_string(step + 1) + ".txt";
            std::ofstream fout2(fileName2);
            for (int i = 0; i < NUMGRID; i++)
                for (int j = 0; j < NUMGRID; j++)
                {
                    fout2 << grids[i][j].direction[0] << " " << grids[i][j].direction[1] << std::endl;
                }
            fout2.close();
        }
    }
}

int main()
{
    double startTime, endTime;
    startTime = omp_get_wtime();

    for (int simID = 1; simID <= NUMSIMS; simID++)
    {
        simulation(simID);
        ANGLE += 2;
        std::cout << "Simulation " << simID << " out of " << NUMSIMS << std::endl;
    }

    endTime = omp_get_wtime();
    std::cout << "Total time: " << endTime - startTime << " s" << std::endl;
}

File: Fiber_Attraction_ABP/Fiber_Attraction_ABP.jl
using Random, Distributions
using DelimitedFiles 
include("Construct_Fiber.jl")

#*************Set Parameters**************
const STEP = 20000  # Number of movement steps
const TICK = 10     #Number of steps between each save file

const LENGTH = 1     #The length of box
const DENSITY = 0.1  # The density of system
const RADIUS = 0.02  #The radius of particles
const DELTATIME = 1  #particles[i].position[0] += DELTATIME * particles[i].velocity[0]
const V = 0.001     #Initial velocity

const DIFFCOEFF = 1         #The diffusion coefficient
const KECM = RADIUS * 0.0001 #The coefficient for ECM force
const ANGLE = 1                  #For ECM force

const PI = 3.1415926
const AREAOFBOX = LENGTH * LENGTH
const NUMBER = floor(Integer, AREAOFBOX * DENSITY / (PI * RADIUS * RADIUS))

#*************Add environment effect
const NUMGRID = 5     # Number of gird
const fiberChangeCoeff = 0.2  #for direction of fiber
fiberRecoverCoeff = 0 # for direction of fiber recover
const fiberEffectCoeff = 0  #for velocity of particles

#***********reconstruction of fiber
sizeBin = 0.5  # deltaR = sizeBin * r
rStepCoeff = 0.25 # rstep = initr * rStepCoeff
alpha = -20 # theoretical = exp.(alpha * correlationX)
const numOfRun = 10000 




#***********Random Parameters
normal = Normal(0, PI/64)
uniform = Uniform(0, 1)

mutable struct Grid
    direction :: Array{Float64,1}
    numOfParticles ::Int64
    particlesInGrid ::Array{Int64,1}
    position :: Array{Float64,1}
    initDirection :: Array{Float64,1}
end

mutable struct Particle
     position :: Array{Float64,1}
     velocity :: Array{Float64,1}
     labelofgrid :: Array{Int64,1}
     force :: Array{Float64,1}
     correctMove :: Array{Float64,1}
end

function boundaryDisplacement(a, b)
    delta = a - b
    if abs(delta) > LENGTH / 2
        if delta > 0
            delta -= LENGTH
        else
            delta += LENGTH
        end
    end
    return delta
end

function boundaryDistance(a, b)
    deltax = boundaryDisplacement(a.position[1], b.position[1])
    deltay = boundaryDisplacement(a.position[2], b.position[2])

    return sqrt(deltax * deltax + deltay * deltay)
end

function boundaryCondition(particle)
    if particle.position[1] >= LENGTH
        particle.position[1] = particle.position[1] - LENGTH
    end
    if particle.position[1] < 0
        particle.position[1] = particle.position[1] + LENGTH
    end
    if particle.position[2] >= LENGTH
        particle.position[2] = particle.position[2] - LENGTH
    end
    if particle.position[2] < 0
        particle.position[2] = particle.position[2] + LENGTH
    end
    return particle.position
end

function isOverlap(a, b)
    return (boundaryDistance(a, b) < 2 * RADIUS)
end

function isOppositeMove(particleA, particleB)
    d = boundaryDistance(particleA, particleB);
    dv = sqrt(particleA.velocity[1] * particleA.velocity[1] + particleA.velocity[2] * particleA.velocity[2])

    deltax = boundaryDisplacement(particleA.position[1], particleB.position[1])
    deltay = boundaryDisplacement(particleA.position[2], particleB.position[2])

    cosTheta = (-deltax * particleA.velocity[1] + (-deltay) * particleA.velocity[2]) / (d * dv)

    return (cosTheta >= cos(ANGLE * PI / 180) && (particleA.velocity[1] * particleB.velocity[1] + particleA.velocity[2] * particleB.velocity[2] < 0))
end

function initializationGrids()
    grids = Array{Grid}(undef, NUMGRID,NUMGRID)
    Threads.@threads for i = 1:NUMGRID
        for j = 1:NUMGRID
            grids[i,j] = Grid([cos(2*PI*rand(uniform)),sin(2*PI*rand(uniform))],0,[],[(i-1)*(LENGTH / NUMGRID) + LENGTH / NUMGRID / 2, (j-1)*(LENGTH / NUMGRID) + LENGTH / NUMGRID / 2],[1,0])
        end
    end

    return grids
end

function initializationParticles()
    particles= Array{Particle}(undef, NUMBER)
    Threads.@threads for ii = 1:NUMBER
        particles[ii] = Particle([0,0],[0,0],[0,0],[0,0],[0,0])
    end
    i = 1
    while i <= NUMBER
        particles[i].position = [LENGTH * rand(uniform) , LENGTH * rand(uniform)] 
        particles[i].labelofgrid = [min(ceil(Integer, particles[i].position[1] / (LENGTH / NUMGRID)), NUMGRID) , min(ceil(Integer, particles[i].position[2] / (LENGTH / NUMGRID)), NUMGRID)]
  
        hasOverlap = false
        for j = 1:i
            if i != j && isOverlap(particles[i], particles[j])
                hasOverlap = true
                break
            end
        end
        if !hasOverlap
            i += 1
        end
    end
#pragma omp parallel for private(ii)
    Threads.@threads for ii = 1:NUMBER
        particles[ii].velocity = [V * LENGTH * (rand(uniform)-0.5) , V * LENGTH * (rand(uniform)-0.5)]
    end
    return particles
end

function rotateAngleAtoB(vectorA::Array, vectorB::Array, coeff::Number)
    vectorAangle = atan(vectorA[2], vectorA[1])
    vectorBangle = atan(vectorB[2], vectorB[1])
    diff = vectorBangle - vectorAangle
    sign = (diff < 0) ? -1.0 : 1.0
    while abs(diff) > PI / 2
        diff -= PI * sign
    end
    return diff * coeff
end

function rotateAtoB(vectorA::Array, vectorB::Array, coeff::Number)
    absVectorA = sqrt(vectorA[1] * vectorA[1] + vectorA[2] * vectorA[2])
    vectorAangle = atan(vectorA[2], vectorA[1])
    vectorAangle += rotateAngleAtoB(vectorA, vectorB, coeff)
    vectorA = [absVectorA * cos(vectorAangle),absVectorA * sin(vectorAangle)]
    return vectorA
end

function rotateAtoB(vectorA::Array, rotateAngle::Number)
    absVectorA = sqrt(vectorA[1] * vectorA[1] + vectorA[2] * vectorA[2])
    vectorAangle = atan(vectorA[2], vectorA[1])
    vectorAangle += rotateAngle 
    vectorA = [absVectorA * cos(vectorAangle),absVectorA * sin(vectorAangle)]
    return vectorA
end


function fiberAffected(particles, grids)
#pragma omp parallel for private(i)

    #Threads.@threads 
    for i = 1: NUMBER
        particles[i].labelofgrid = [min(ceil(Integer, particles[i].position[1] / (LENGTH / NUMGRID)), NUMGRID) , min(ceil(Integer, particles[i].position[2] / (LENGTH / NUMGRID)), NUMGRID)]

        grids[particles[i].labelofgrid[1], particles[i].labelofgrid[2]].numOfParticles +=1 
        append!(grids[particles[i].labelofgrid[1], particles[i].labelofgrid[2]].particlesInGrid, i)  
    end

    #pragma omp parallel for private(ii)
    #Threads.@threads 
    for i = 1:NUMGRID
        for j = 1:NUMGRID
            rotateAngle = 0
            if (grids[i, j].numOfParticles >= 2)
                for p1 = 1:grids[i, j].numOfParticles
                     for p2 = p1 + 1 :  grids[i, j].numOfParticles
                         a = grids[i, j].particlesInGrid[p1]
                         b = grids[i, j].particlesInGrid[p2]
                         if (isOppositeMove(particles[a], particles[b]))
                            rotateAngle += rotateAngleAtoB(grids[i, j].direction, particles[a].position - particles[b].position, fiberChangeCoeff)
                         end
                    end
                end
                grids[i, j].direction = rotateAtoB(grids[i, j].direction, rotateAngle)    
            else
                grids[i, j].direction = rotateAtoB(grids[i, j].direction, grids[i, j].initDirection, fiberRecoverCoeff);
            end
            grids[i, j].numOfParticles = 0
        end
    end
end

function moveParticles(particles, grids)
    # Move all particles. Update the positions of particles 
    # according to the current velocities.Velocities are updated 
    # by Gaussian random walk. Gaussian distribution is set by
    # global normal distribution generator.

    #pragma omp parallel for private(i)
    Threads.@threads for i = 1:NUMBER 
        particles[i].labelofgrid = [min(ceil(Integer, particles[i].position[1] / (LENGTH / NUMGRID)), NUMGRID) , min(ceil(Integer, particles[i].position[2] / (LENGTH / NUMGRID)), NUMGRID)]
        #Add force
        particles[i].velocity[1] += particles[i].force[1]
        particles[i].velocity[2] += particles[i].force[2]
        #environment effect
        particles[i].velocity = rotateAtoB(particles[i].velocity, grids[particles[i].labelofgrid[1], particles[i].labelofgrid[2]].direction, fiberEffectCoeff)
        #rotation diffusion
        particles[i].velocity = rotateAtoB(particles[i].velocity, rand(normal))
        #Update positions
        particles[i].position[1] += DELTATIME * particles[i].velocity[1];
        particles[i].position[2] += DELTATIME * particles[i].velocity[2];
        # Correct the position according to periodic boundary condition.
        particles[i].position = boundaryCondition(particles[i])
    end
end

function correctOverlap(particleA, particleB)
    
    # Offset overlap between two particles.If two particles overlap, 
    # move each one-half the overlap distance along their center-to-center axis.
    
    d = boundaryDistance(particleA, particleB)
    deltax = boundaryDisplacement(particleA.position[1], particleB.position[1])
    deltay = boundaryDisplacement(particleA.position[2], particleB.position[2])
    deltaD = 2 * RADIUS - d

    particleA.correctMove[1] += 0.5 * deltaD * deltax / d
    particleA.correctMove[2] += 0.5 * deltaD * deltay / d

end

function recordCorrectOverlap(particles)
#pragma omp parallel for private(j)
    Threads.@threads for i = 1:NUMBER
        for j = 1: NUMBER
            if j != i
                if (isOverlap(particles[i], particles[j]))
                    correctOverlap(particles[i], particles[j])
                end
            end
        end
    end
end

function moveCorrectOverlap(particles)
#pragma omp parallel for private(i)
    recordCorrectOverlap(particles)
    for i = 1:NUMBER
        particles[i].position[1] += particles[i].correctMove[1]
        particles[i].position[2] += particles[i].correctMove[2]
        particles[i].correctMove = [0, 0]
        particles[i].position = boundaryCondition(particles[i])
    end
end


function simulation(simID)
    # Run a complete simulation from start. simID labels the save detination for data.
    # After initialization, particles move for a number of steps specified by the global
    # constant STEP. Motion data is saved for every TICK steps.
    grids = initializationGrids()
    generateFiber(grids,sizeBin, rStepCoeff,alpha,numOfRun)
    particles = initializationParticles()
    
    positiondata = Array{Float64}(undef, NUMBER, 4)
    fiberPositiondata = Array{Float64}(undef, NUMGRID*NUMGRID,2)
    for i = 1:NUMBER
        positiondata[i,:] = [particles[i].position[1], particles[i].position[2], particles[i].velocity[1], particles[i].velocity[2]]
    end
    writedlm("RecoverCoeff$(round(fiberRecoverCoeff; digits=1))/position0.csv", positiondata, ',')
    k = 1
    for i = 1:NUMGRID
        for j = 1:NUMGRID
            fiberPositiondata[k,:] =[grids[i,j].direction[1],grids[i,j].direction[2]]
            k +=1 
        end
    end
    writedlm("RecoverCoeff$(round(fiberRecoverCoeff; digits=1))/fiberPosition0.csv",  fiberPositiondata, ',')

    for step = 1:STEP
        for i = 1:NUMBER
            particles[i].force=[0,0]
        end
        moveParticles(particles, grids)
        moveCorrectOverlap(particles)
        fiberAffected(particles, grids)

        if step % TICK == 0
            for i = 1:NUMBER
                positiondata[i,:] = [particles[i].position[1], particles[i].position[2], particles[i].velocity[1], particles[i].velocity[2]]
            end
            writedlm("RecoverCoeff$(round(fiberRecoverCoeff; digits=1))/position$(step).csv", positiondata, ',')

            k = 1
            for i = 1:NUMGRID
                for j = 1:NUMGRID
                    fiberPositiondata[k,:] =[grids[i,j].direction[1],grids[i,j].direction[2]]
                    k +=1 
                end
            end
            writedlm("RecoverCoeff$(round(fiberRecoverCoeff; digits=1))/fiberPosition$(step).csv",  fiberPositiondata, ',')
        end
    end 
end

function runSimulation()
    for i = 1:6
        println("simulation: ", i)
        simulation(i)
        global fiberRecoverCoeff += 0.2
    end
end

@time runSimulation()


